src/params_loader.rs

```diff
@@ -15,9 +17,82 @@ pub trait PipesParamsLoader {
     /// to create a PipesContext or should instead return a mock.
     fn is_dagster_pipes_process(&self) -> bool;
     /// Load params passed by the orchestration-side context injector.
-    fn load_context_params(&self) -> Map<String, Value>;
+    fn load_context_params(&self) -> Result<Map<String, Value>, ParamsError>;
     /// Load params passed by the orchestration-side message reader.
-    fn load_message_params(&self) -> Map<String, Value>;
+    fn load_message_params(&self) -> Result<Map<String, Value>, ParamsError>;
+}
+
+#[derive(Debug, Error)]
+#[error("{} parameter \"{}\" is {}", .origin, .param, .source)]
+pub struct ParamsError {
+    pub param: String,
+    pub origin: ParamOrigin,
+    pub source: ParamsErrorKind,
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ParamOrigin {
+    Cli,
+    EnvVar,
+}
+
+impl fmt::Display for ParamOrigin {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Cli => write!(f, "cli"),
+            Self::EnvVar => write!(f, "env var"),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum ParamsErrorKind {
+    #[error("not present")]
+    #[non_exhaustive]
+    NotPresent,
+
+    #[error("invalid")]
+    #[non_exhaustive]
+    Invalid {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+}
+
+impl From<std::env::VarError> for ParamsErrorKind {
+    fn from(value: std::env::VarError) -> Self {
+        match value {
+            std::env::VarError::NotPresent => Self::NotPresent,
+            std::env::VarError::NotUnicode(_) => Self::Invalid {
+                source: Box::new(value),
+            },
+        }
+    }
+}
+
+impl From<base64::DecodeError> for ParamsErrorKind {
+    fn from(value: base64::DecodeError) -> Self {
+        Self::Invalid {
+            source: Box::new(value),
+        }
+    }
+}
+
+impl From<std::io::Error> for ParamsErrorKind {
+    fn from(value: std::io::Error) -> Self {
+        Self::Invalid {
+            source: Box::new(value),
+        }
+    }
+}
+
+impl From<serde_json::Error> for ParamsErrorKind {
+    fn from(value: serde_json::Error) -> Self {
+        Self::Invalid {
+            source: Box::new(value),
+        }
+    }
```

Comment:
Mapping all lower-level errors to `ParamsErrorKind::Invalid`. 

I'm doing this manually as I haven't found a way to auto-derive `From` for multiple error types using `thiserror`.